module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,334 @@
+     * @return object
+     * @throws AuthException
+     */
+    protected function getProvider(): object
+    {
+        if (!isset($this->provider)) {
+            $url = $this->getConfig()->OpenIDConnect->url;
+            $url .= str_ends_with($url, '/') ? '' : '/';
+            $url .= '.well-known/openid-configuration';
+            try {
+                $this->provider = json_decode($this->httpService->get($url)->getBody());
+            } catch (\Exception $e) {
+                throw new AuthException(
+                    'Cannot fetch provider configuration: ' . $e->getMessage()
+                );
+            }
+        }
+        return $this->provider;
+    }
+
+    /**
+     * Validate configuration parameters. This is a support method for getConfig(), so the configuration MUST be
+     * accessed using $this->oidcConfig; do not call $this->getConfig() from within this method!
+     *
+     * @throws AuthException
+     * @return void
+     */
+    protected function validateConfig(): void
+    {
+        $requiredParams = ['url', 'client_id', 'client_secret'];
+        foreach ($requiredParams as $param) {
+            if (empty($this->oidcConfig?->OpenIDConnect?->$param ?? null)) {
+                throw new AuthException(
+                    'One or more OpenID Connect parameters are missing. Check your OpenIDConnect.ini!'
+                );
+            }
+        }
+    }
+
+    /**
+     * Attempt to authenticate the current user. Throws exception if login fails.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing account credentials.
+     *
+     * @throws AuthException
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    public function authenticate($request): \VuFind\Db\Row\User
+    {
+        $code = $request->getQuery()->get('code');
+
+        if (empty($code)) {
+            throw new AuthException('authentication_error_admin');
+        }
+        $request_token = $this->getRequestToken($code);
+        $state = $request->getQuery()->get('state');
+        $stateIsValid = $state == $this->session->oidc_state;
+        unset($this->session->oidc_state);
+        if (!$stateIsValid) {
+            throw new AuthException('authentication_error_admin: bad state');
+        }
+
+        $claims = $this->decodeJWT($request_token->id_token, 1);
+
+        if (!$this->validateIssuer($claims->iss)) {
+            throw new AuthException('authentication_error_admin: wrong issuer');
+        }
+        $claimsValid = $this->verifyJwtClaims($claims);
+        unset($this->session->oidc_nonce);
+        if (!$claimsValid) {
+            throw new AuthException('authentication_error: not valid claims');
+        }
+
+        $access_token = $request_token->access_token;
+
+        $user_info = $this->getUserInfo($access_token);
+        $user = $this->getUserTable()->getByUsername($user_info->sub);
+        if (isset($user_info->given_name)) {
```

Comment:
Is it worth making attribute mapping configurable like in [Shibboleth](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Auth/Shibboleth.php#L223-L240), or is the Open ID Connect spec so limited that there's no possibility of values beyond what is covered here?

Reply:
There is no need to make id configurable, attributes are defined in specification: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

Reply:
Is there a possibility of non-standard claims being included? e.g. might somebody be able to pass catalog credentials through this mechanism, and if so, would we want to allow those to be mapped here?

Reply:
That certainly seems plausible to me. The other side could be e.g. a VuFind instance that supports custom mappings with the OIDC provider. :)

Reply:
It does make sense, i added basic configuration for attributes mapping

Reply:
Is it worth adding special handling for cat_username/cat_password similar to [the CAS driver's approach](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Auth/CAS.php#L146-L172) just in case? Or do you prefer to leave that out until it is needed? If leaving it out, maybe we should add a comment indicating unsupported fields, or throw an exception in the driver.

Reply:
I updated the code to filter available attributes and handle cat_username/cat_password like it is done in CAS driver